### ISSUE TITLE
API Defect List (to be corrected for API v4)

### ISSUE THREAD

**Issue by myrsloik** [vapoursynth/vapoursynth#59]
~~Add major API version to the 3 public header files:~~
~~vapoursynth4.h~~
~~vshelper4.h~~
~~vsscript4.h~~

~~Add getAPIVersion() to vsapi. Will simplify a lot of things.~~

~~A createCore function where you can prevent autoloading of plugins.~~

~~Move plugin identifiers to vshelper and adopt consistent naming of things~~

~~Don't implement scanline based filters but adjust the API so it's possible to do later. For example require explicit declaration of node dependencies.~~

Additional Enums #570 

Make it possible to cancel in progress requests to getFrameAsync? Return a proper future/promise style object instead?

Add automatic packing/unpacking to avs compat since compat formats are now removed. Easiest done by introducing packing/unpacking filters.

Improve plugin handling on windows to leak less memory

The possibility to register user functions with argument lists and export them through the core like a plugin. Add a user function namespace as an experiment?

~~Graph inspection - also add a processing time field that can be retrieved~~

~~Per core message handler~~

~~SUPPORT REMOVED since nobody uses this feature. It's really awkward to write filters that output multiple clips. Or more exactly kinda weird when both frames (such as main frame and alpha) are generated at the same time and therefore pushing both at once would be very helpful for performance.~~

~~All "void **instanceData" in filters should be changed to "void *instanceData". If extra indirection is needed it's trivial to do so at creation time. Zero filters have benefited from the extra indirection so far.~~

~~The "void **frameData" argument in VSFilterGetFrame  should be changed to void *[4] to make it trivial to store a pointer (to a node usually) and an offset without allocation. Will improve splice getframe performance when a temporary allocation is no longer needed (or a complex recalculation of decisions).~~

~~Require filters to set a flag to receive arFrameReady events.~~

~~Allow a function to declare that it accepts everything and leave the handling to it~~
~~Allow a function to declare that it can return anything~~

~~Fix the weird cm prefix of colorfamily, should obviously be cf.~~

~~Remove ycocg as a colorfamily. Nobody uses it and it belongs to the YUV family anyway~~

~~registerFunction should return success or failure (function already exists, uses invalid identifier or other problem)~~

~~Review plugin loading in general. The current model makes changes extremely hard with the few things passed to registerFunction.~~

~~callFunc(VSFuncRef *func, const VSMap *in, VSMap *out, VSCore *core, const VSAPI *vsapi) should be callFunc(VSFuncRef *func, const VSMap *in, VSMap *out)~~

~~Stride should be ptrdiff_t.~~ All other types in the API should be reviewed as well?

~~createFilter and filterInit overlap a lot. This means that filterInit usually ends up only setting videoinfo and nothing else. Allow videoinfo to be passed to createFilter instead and save some boilerplate code.~~

~~readOnly in configPlugin should be renamed to flags and formalized.~~

~~Prefix all functions in vshelper4.h with vsh_ or something similar for C. Maybe add a vsh namespace for C++ mode~~

~~Rename vs_normalizeRational to vs_reduceRational to make more sense~~

~~Enumerating all plugins and their ids/namespaces/whatever is kinda messy. Improve it.~~

~~queryCompletedFrame is half-broken. Fix it.~~

~~Remove getFormat and friends. Replace with a simple int that codes all the relevant data. Do the same for audio.~~

.~~propGetInt/Float should have a version to get a saturated int or float instead of only int64_t or double as the returned type. Will save plugins writers a lot casting without losing the internal representation.~~

~~Type hint for binary/utf8 for the vsmap data type make inspection filters behave better and not print junk.~~

~~Fix the vsscript api so it actually makes sense.~~

~~Add some way to set core creation flags from python and vsscript.~~

~~Change the vsscript api to export a single struct with function pointers as well.~~

~~Plugin return type hints~~ 

~~#601 Plugin version API~~ 

~~Remove compat formats since mvtools is now fully native.~~

~~Add map* functions that consume references~~

~~cloneFrame/Node/FunctionRef should be renamed to addFrame/Node/FunctionRef~~

~~Drop Ref from VSFrameRef, VSNodeRef and VSFunctionRef since it's pointless indirection~~

~~Remove multiple output nodes support completely even for api3 to reduce complexity more~~

Ideas skipped this time:
Filters cannot simply yield their execution time and return to the end of the execution queue. This could be useful for filters that transfer to/from GPU memory. This feature may not be added or be saved for a later API revision. Or should filters simply be able to reserve/release execution threads like is done internally already?
Support for passing frames stored in OpenCL memory
Tiling/line based filters for more effective cache usage

~~Also kinda depends on #508~~

**Comment by chainikdn** [comment]
Please add properties to the nodes: 
VSMap* getNodeProps(VSNodeRef*)

Currently there's no straight way to pass something between nodes at create/init time.
For example MVTools build for Vapoursynth currently uses really ugly solution with passing some values via frame #0 properties (and requesting frame #0 from the filter init function).

**Comment by myrsloik** [comment]
I guess I should add that somehow

**Comment by chainikdn** [comment]
Great!
And another one: a method to get script variable could be useful (something like IScriptEnvironment::GetVar())

**Comment by dubhater** [comment]
I plan to move the initialisation to the getFrame functions, so it won't be so ugly anymore.

**Comment by myrsloik** [comment]
This isn't avisynth, use frame properties. Global variables will never work properly the way you want them to

**Comment by chainikdn** [comment]
OK, let's talk about mpv integration then ;). Seems like the only way to get source video frame rate is to use global variable "container_fps".

**Comment by chainikdn** [comment]
> I plan to move the initialisation to the getFrame functions, so it won't be so ugly anymore.

here's another workaround:

`void create(const VSMap *in, VSMap *out, ...) {
`    vsapi->propSetInt(out,"value",123456,0);
`    vsapi->createFilter(in,out,...)
`}

`A = Filter1(clip)
`B = Filter2(A["clip"],A["value"])

**Comment by dubhater** [comment]
Eh, there's no need to break every script ever. It's perfectly doable with a mutex.

**Comment by dubhater** [comment]
A simple way to obtain container_fps:
```python
video_in = core.std.AssumeFPS(video_in, fpsnum=int(container_fps * 1000), fpsden=1000)
```

**Comment by myrsloik** [comment]
`A = Filter1(clip)
B = Filter2(*A)`
This would be a more natural way to do it in python and already works I guess. But whatever. It works well enough now and there's only one filter doing it. And that's because it copied the original avisynth way.

The fact that you can do it this way already is the main reason I didn't add clip properties. You can simply stuff them all into the return map and then it's the scripting language's problem.

**Comment by dubhater** [comment]
Maybe propGetInt and similar functions could take a default value instead of error. There are two other ways to check if a key exists.

**Comment by Bubbymathis** [comment]
Trash

**Comment by Endilll** [comment]
I want to put in a word for compat formats, specifically COMPATBGR32, because without it rendering frame via GUI framework like Qt or WxWidgets would become much more pain. For instance, VSEdit relies on it.

I think removing it as a format and adding some API instead would do the work, too, if there are reasons not to keep it.

**Comment by myrsloik** [comment]
The compat formats will probably stay for quite a while since keeping them overall creates very little extra work. This is more like a 5+ year plan.

**Comment by IFeelBloated** [comment]
> Find one case where copyMap() would be useful and I'll add it. VSMap is simply meant to pass data either in or out so there's no real reason for a copyMap() to exist.

https://github.com/IFeelBloated/vapoursynth-mvtools-sf/blob/master/src/MVAnalyze.hxx#L469

**Comment by stuxcrystal** [comment]
While you're implementing APIv4.
Perhaps it would be a good idea to somehow let plugins specify what they return this would solve problems with function-signature generation in Python.

**Comment by myrsloik** [comment]
Makes sense. I guess.

> While you're implementing APIv4:
> Perhaps it would be a good idea to somehow let plugins specify what they return in each function. This would solve problems with function-signature generation in Python.
> 
> For example, vsgenstubs currently just assumes every plugin-function just returns clips which is not the case.

**Comment by myrsloik** [comment]
I also want to have a better way to export function arguments in an already parsed way. The current stuff is so horrible.

**Comment by stuxcrystal** [comment]
I think `const VSPLUGINAPI *vspapi` should be only be able to be obtained through the existing VapourSynth-API:
That way you can version the `VSPLUGINAPI*` struct and change it if neccessary.

I feel something like `const VSPLUGINAPI* getPluginAPI(VSPlugin*)` should suffice.

**Comment by myrsloik** [comment]
It can't be implemented that way. Mostly because which major version of the VSAPI struct to provide isn't known until configPlugin has been called (assuming you don't simply bump the entry point number by one every time things change which would instead make it implicit). I doubt there are any more changes needed in the future and the entry point name can always be bumped as well.

**Comment by myrsloik** [comment]
> While you're implementing APIv4:
> Perhaps it would be a good idea to somehow let plugins specify what they return in each function. This would solve problems with function-signature generation in Python.
> 
> For example, vsgenstubs currently just assumes every plugin-function just returns clips which is not the case.

Since we're on the subject of arguments. My plan at the moment is to have a way to enumerate all functions in a plugin (that vsmap string dump is horrible) and getting the plugin part of the info as well. Is returning the in and out arguments as a string good enough? Should that be improved too?

**Comment by OrangeChannel** [comment]
Not sure if it just hasn't been pushed to doodle1 yet with a bunch of other documentation changes, but `vsscript_getOutput2` introduced in https://github.com/vapoursynth/vapoursynth/commit/97caf8b6acb6942ffeaeaeb939aa54925f630675 isn't documented.

**Comment by myrsloik** [comment]
> Not sure if it just hasn't been pushed to doodle1 yet with a bunch of other documentation changes, but `vsscript_getOutput2` introduced in [97caf8b](https://github.com/vapoursynth/vapoursynth/commit/97caf8b6acb6942ffeaeaeb939aa54925f630675) isn't documented.

Documentation isn't done at all so don't bother reporting it.
